### PR TITLE
[WIP] ManageIQ Utils: moving all shared code to manageiq_utils

### PR DESCRIPTION
**Pull request by dkorn**
The utils module is a best practice/guideline of modules part of the `ansible/ansible` repo.

In addition:
* making the required changes in manageiq_user module
* adding the GNU General Public License V3 to the files headers
(separate PR#60 takes care of the COPYING file)

TODO (dkorn): Fix tests

### CODE REVIEW

File: library/manageiq_user.py [hunk 1]
```diff
@@ -83,180 +101,140 @@
...
 
-        user_id = self.find_user_by_userid(userid)
-        if user_id:  # user already exist
-            return self.update_user_if_required(user_id, userid, username, group_id, password, email)
+        user = manageiq.find_collection_resource_by('users', userid=userid)
+        if user:  # user already exist
+            return self.update_user_if_required(manageiq, user, userid, username, group['id'], password, email)
         else:
-            return self.create_user(userid, username, group_id, password, email)
+            return self.create_user(manageiq, userid, username, group['id'], password, email)
```
Comment (cben) [inline library/manageiq_user.py]: I'd maybe give `manageiq` to the constructor, and use `self.manageiq` everywhere.
That way you don't have to pass it through from method to method.

File: library/manageiq_user.py [hunk 2]
```diff
@@ -83,180 +101,140 @@
...
+        user = manageiq.find_collection_resource_by('users', userid=userid)
+        if not user:  # user doesn't exist
             return dict(
                 changed=self.changed,
                 msg="User {userid} does not exist in manageiq".format(userid=userid))
         try:
-            url = '{api_url}/users/{user_id}'.format(api_url=self.api_url, user_id=user_id)
-            result = self.client.post(url, action='delete')
+            url = '{api_url}/users/{user_id}'.format(api_url=manageiq.api_url, user_id=user['id'])
+            result = manageiq.client.post(url, action='delete')
```
Comment (cben) [inline library/manageiq_user.py]: I think it's worth having `manageiq.post` wrapper that takes a relative url like `users/{id}` and prepends `api_url` automatically.

File: library/manageiq_user.py [hunk 3]
```diff
@@ -83,180 +101,140 @@
...
         Returns:
             the created user id, name, created_on timestamp,
             updated_on timestamp, userid and current_group_id
         """
-        if not self.user_update_required(user_id, userid, username, group_id, email):
+        if not self.user_update_required(user, username, group_id, email):
             return dict(
                 changed=self.changed,
                 msg="User {userid} already exist, no need for updates".format(userid=userid))
+        url = '{api_url}/users/{user_id}'.format(api_url=self.api_url, user_id=user.id)
```
Comment (cben) [inline library/manageiq_user.py]: s/self.api_url/manageiq.api_url/ ?

**Review by cben — COMMENTED** [review]
Very nice!  The signal-to-noise in the module is much better now.  Few small suggestions.

Cross-ref upstream PR: https://github.com/ansible/ansible/pull/25106

File: library/manageiq_user.py (hunk 2 above)
Comment (dkorn) [reply library/manageiq_user.py]: I agree, but I prefer to add it in a separate PR, this one is quite a big change already.

File: library/manageiq_user.py (hunk 3 above)
Comment (dkorn) [reply library/manageiq_user.py]: Done, thanks!

File: library/manageiq_user.py (hunk 1 above)
Comment (dkorn) [reply library/manageiq_user.py]: Interesting suggestion, I'll give it a try

File: library/manageiq_user.py (hunk 3 above)
Comment (cben) [reply library/manageiq_user.py]: didn't push?

Comment (dkorn) [reply library/manageiq_user.py]: @cben soon :)

File: library/manageiq_utils.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    """
+
+    def __init__(self, module):
+        """
+            module         - an AnsibleModule instance
+            url            - manageiq environment url
+            user           - the username in manageiq
+            password       - the user password in manageiq
+            miq_verify_ssl - whether SSL certificates should be verified for HTTPS requests
+            ca_bundle_path - the path to a CA_BUNDLE file or directory with certificates
```
Comment (cben) [inline library/manageiq_utils.py]: is this is a remainder from earlier version where this constructor took all these params separately?

File: setup.py [hunk 5]
```diff
@@ -9,6 +9,6 @@
     package_dir={'': 'library'},
     py_modules=["manageiq_provider", "manageiq_policy_assignment",
                 "manageiq_custom_attributes", "manageiq_user",
-                "manageiq_tag_assignment"],
+                "manageiq_tag_assignment", "manageiq_utils"],
```
Comment (cben) [inline setup.py:12]: Tests fail on `import manageiq as manageiq_utils`.  Not sure why, but perhaps dropping the `_utils` here will help?

**Comment by cben** [comment]
A similar refactoring has been done as the modules were rewritten into Ansible itself:
https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/manageiq.py

We significantly rewrote all modules anyway, with incompatible interface changes, so we didn't even try to keep this repo in sync.  Closing.

**Pull request closed without merging**